adapters/eplanning/eplanning.go

```diff
@@ -232,13 +286,28 @@ func verifyImp(imp *openrtb.Imp) (*openrtb_ext.ExtImpEPlanning, error) {
 	return &impExt, nil
 }
 
-func getSizeFromImp(imp *openrtb.Imp) (uint64, uint64) {
+func indexOf(element string, data []string) int {
+	for k, v := range data {
+		if element == v {
+			return k
+		}
+	}
+	return -1
+}
+
+func getSizeFromImp(imp *openrtb.Imp, isMobile bool) (uint64, uint64) {
 	if imp.Banner.W != nil && imp.Banner.H != nil {
 		return *imp.Banner.W, *imp.Banner.H
 	}
 
 	if imp.Banner.Format != nil {
-		for _, format := range imp.Banner.Format {
+		sizesSortedByPriority := imp.Banner.Format
+		if isMobile {
+			sort.Sort(byPriorityMobile(sizesSortedByPriority))
+		} else {
+			sort.Sort(byPriorityDesktop(sizesSortedByPriority))
+		}
+		for _, format := range sizesSortedByPriority {
 			if format.W != 0 && format.H != 0 {
 				return format.W, format.H
```

Comment:
If at the end, only one format is going to be included in the `ImpExt`, do we need to sort the entire list? Assuming the format list comes with N entries, sorting comes at a cost of O(NlogN). But, if we just compare the format list and select a match from the `priorityOrderFor<DEVICE>Sizes` list, we could, at the cost of a little memory that Prebid Server could spare, do this in O(N+K) time instead. This seems like a good optimization opportunity given that the sorting algorithm you are currently using swaps to the end only the `priorityOrder` matches (see test case number two of the `TestLessByPriorityMobile(t *testing.T)` I wrote).

Maybe we could do something similar to:

```
298   func getSizeFromImp(imp *openrtb.Imp, isMobile bool) (uint64, uint64) {
299       if imp.Banner.W != nil && imp.Banner.H != nil {
300           return *imp.Banner.W, *imp.Banner.H
301       }
302  
303       if imp.Banner.Format != nil {
304 -         sizesSortedByPriority := imp.Banner.Format
305 -         if isMobile {
306 -             sort.Sort(byPriorityMobile(sizesSortedByPriority))
307 -         } else {
308 -             sort.Sort(byPriorityDesktop(sizesSortedByPriority))
309 -         }
310 -         for _, format := range sizesSortedByPriority {
311 -             if format.W != 0 && format.H != 0 {
312 -                 return format.W, format.H
313 -             }
314 -         }
    +         // Space O(n)
    +         hashedFormats := make(map[string]int,0)
    +
    +         // Traverse format array in O(n) time
    +         for i, format := range imp.Banner.Format {
    +             if format.W != 0 && format.H != 0 {
    +                 hashedFormats[fmt.Sprintf("%dx%d", format.W, format.H)] = i
    +             }
    +         }
    +         // Traverse prioritized formats in O(k) time
    +         for i:=len(priorityOrderFor__MobileOrDesktop__SizesAsc)-1; i>=0; i-- {
    +             //   ^--- Start with the highest priority in priorityOrderFor__MobileOrDesktop__SizesAsc
    +
    +             if formatIndex, wasFound := hashedFormats[priorityOrderFor__MobileOrDesktop__SizesAsc]; wasFound {
    +                 return imp.Banner.Format[formatIndex].W, imp.Banner.Format[formatIndex].H
    +             }
    +         }
315       }
316  
317       return 0, 0
318   }
adapters/eplanning/eplanning.go
```